runtime/compiler/control/JitDump.cpp

```diff
@@ -515,6 +516,13 @@ runJitdump(char *label, J9RASdumpContext *context, J9RASdumpAgent *agent)
    trfflush(jitdumpFile);
    trfclose(jitdumpFile);
 
+   recompilationThreadInfo->suspendCompilationThread();
+   while (recompilationThreadInfo->getCompilationThreadState() != COMPTHREAD_SUSPENDED)
+      {
+      //compInfo->getCompilationMonitor()->notifyAll();
+      //compInfo->waitOnCompMonitor(recompilationThreadInfo->getCompilationThread());
+      }
+
    compInfo->getPersistentInfo()->setDisableFurtherCompilation(false);
```

Comment:
Converting to draft until this is sorted out. @dsouzai this is not really ideal. See commit message from 3a7af3184ffc2715409da693906ef5eb3e0a6ee0 as to why we want this. I would like to avoid spinning here wasting cycles waiting for the diagnostic thread to suspend. Is there some monitor or something I can wait on instead?

Reply:
Why do we need to wait? What happens if we let the current thread go and the diagnostic will thread will suspend itself in the near future?

Reply:
The reason is because 213d7696458648493ea86867031665188e9c7212 made it so that a diagnostic threads should be the only ones who can process JitDump compile requests. The diagnostic threads also cannot self-suspend, i.e. this line in the same commit above:

https://github.com/eclipse/openj9/blob/726f491e4c0644d49d224f560bb10095bf4aec05/runtime/compiler/control/CompilationThread.cpp#L5240-L5241

The reason we want to avoid self-suspend for diagnostic threads is because there is a timing hole between when the diagnostic thread is resumed in JitDump.cpp and when the actual JitDump compile request is added to the queue. I've encountered scenarios where within this timing hole, the diagnostic thread self-suspends (prior to 213d7696458648493ea86867031665188e9c7212) and we end up with normal compilation threads picking up JitDump method compile requests, which is definitely something we don't want. See #11772 for details of how that occured.

Reply:
I suppose an alternative is to remove this assert:

https://github.com/eclipse/openj9/blob/726f491e4c0644d49d224f560bb10095bf4aec05/runtime/compiler/control/CompilationThread.cpp#L5248-L5249

And allow diagnostic threads to handle normal compiles. Then we could get away with not waiting in the code being reviewed above. However to me this is not an ideal solution as there are "special" things about the diagnostic thread.

Reply:
Any thoughts on the above?

Reply:
@mpirvu the problem is line 526 in the code being reviewed, that is:

```
   recompilationThreadInfo->suspendCompilationThread();
   while (recompilationThreadInfo->getCompilationThreadState() != COMPTHREAD_SUSPENDED)
      {
      //compInfo->getCompilationMonitor()->notifyAll();
      //compInfo->waitOnCompMonitor(recompilationThreadInfo->getCompilationThread());
      }

   compInfo->getPersistentInfo()->setDisableFurtherCompilation(false);
   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   Problematic line
```

Once we signal the diagnostic thread to suspend, we `setDisableFurtherCompilation(false)`, meaning that other application threads can start queuing compiles. What happens then is that there is a timing hole between the point we ask for the diagnostic thread to suspend and when it actually suspends. Within this timing hole another method may get queued on the compilation queue, and the diagnostic thread may start processing that compile request.

It then hits this assert:
https://github.com/eclipse/openj9/blob/726f491e4c0644d49d224f560bb10095bf4aec05/runtime/compiler/control/CompilationThread.cpp#L5248-L5249

which ensures that diagnostic threads only process JitDump compile requests. Does this make sense? It is also very easy to reproduce this scenario if we don't have that `while` loop above.

Reply:
Looking at the code again, even with the `while` loop above we have a timing hole. The diagnostic thread returns `GO_TO_SLEEP_EMPTY_QUEUE` from `getNextMethodToBeCompiled` if there are no methods on the queue:

https://github.com/eclipse/openj9/blob/726f491e4c0644d49d224f560bb10095bf4aec05/runtime/compiler/control/CompilationThread.cpp#L5238-L5253

But in `processEntries` we change the state of the thread to `COMPTHREAD_WAITING`. This means that  there is a race condition between the JitDump thread signaling the diagnostic thread to suspend and the diagnostic thread executing this code:

https://github.com/eclipse/openj9/blob/726f491e4c0644d49d224f560bb10095bf4aec05/runtime/compiler/control/CompilationThread.cpp#L3999-L4018

We can encounter this situation with two threads:

1. Crashed thread signals diagnostic thread to suspend as per above code being reviewed
2. Meanwhile diagnostic thread is in `processEntries` just about to call `getNextMethodToBeCompiled`
3. `getNextMethodToBeCompiled` returns next action is `GO_TO_SLEEP_EMPTY_QUEUE` because queue is empty
4. We reach `setCompilationThreadState(COMPTHREAD_WAITING);` which overwrites the signal to suspend

This scenario would exist today even without diagnostic threads right? For example if we only have one active compile thread then it will reach this line:

https://github.com/eclipse/openj9/blob/726f491e4c0644d49d224f560bb10095bf4aec05/runtime/compiler/control/CompilationThread.cpp#L5407-L5420

which means if for example the VM singaled us to terminate the thread by changing the state to `COMPTHREAD_SIGNAL_TERMINATE` then this state would get overwritten in the same way as above.

Reply:
@mpirvu thoughts on the above?

Reply:
The sequence of actions I see is the following:
1. crashed thread calls `compileMethod()` to queue a synchronous compilation for the diagnostic thread
2. Diagnostic thread compiles the method and notifies waiting threads (i.e. the crashing thread). At this point the diagnostic thread is in `processEntry()` and holds the compilation queue monitor.
3. Diagnostic thread returns to `processEntries()` and because the stats is still ACTIVE, executes `getNextMethodToBeCompiled()`. Since there is nothing left in the queue, it indicates an action of `GO_TO_SLEEP_EMPTY_QUEYE` and its state changes to `COMPTHREAD_WAITING`. 
4. Diagnostic thread performs a times wait on the compilation queue monitor which releases said monitor.
5. Crashed thread can finally change the state of the diagnostic thread (state can only be changed with compilation queue monitor in hand) to `SIGNAL_SUSPEND`.
6. When new entries are added to the queue the diagnostic thread may be notified, but since its state is no longer `COMPTHREAD_WAITING`, it does nothing, it leaves `processEntries()` and in `run()` it will execute `doSuspend()` which will result in thread suspension.

Reply:
Another idea: since normal compilation threads cannot executed jitdump compilation requests and diagnostic threads cannot execute normal compilation requests, why not use a separate queue for jitdump compilation requests?

Reply:
I updated my branch to the latest state with the latest fixes. I cannot any longer reproduce the assert after calling `suspendCompilationThread` in JitDump.cpp without the `while` loop. I've also verified that in `TR::CompilationInfoPerThread::processEntries()` the compilation monitor is owned by the current thread, meaning that the scenario I described previously is not possible and the scenario outlined by @mpirvu in the second-last comment is correct.

I've updated the PR to simply call `suspendCompilationThread`.